Proposal for groups on socialDB (#2)
NearSocial/SEPs: seps/sep-0002.md

## UX Goals

1. To allow a group of users to moderate with sovereignty a scope of socialDB limited to the members of the group. e.g. group members have read + write access to the group's socialDB namespace. 
1. Non-members only have read-only access to view group content
1. Group members should receive notifications for all relevant social interactions in a group (e.g. Like, Reply for Posts and Discussions)
1. All content posted to groups that a user is part of should appear the user’s "Following" activity feed. 
1. Leaving a group will stop content from appearing on a user’s feed.
1. Group content should only appear in the “All” activity feed if a group member reposts a group post to the main feed.
1. One should be able to retrieve a list of all groups in socialDB
1. One should be able to retrieve a list of all users in a group.
1. One should be able to search for groups, where search text is matched against the group's name and metadata.

> 

Reviewer: Nice these requirements are neat and to us it's mostly sufficient. The only thing I saw is missing for DevHub's use case is that: we allow a content to be post both globally and for a group. Under this model, this will be allow post to both user and to group

Author: We could support that ability without modification to this SEP by 'reposting' a group content to the main post index.  This may give the desired affect, but would result in two conversation threads: one where anyone can read/reply to the main index post and the other group post where only NFT holders can reply but anyone can view it on the group's page.

Reviewer: Repost will work, but two conversation thread seems not ideal. The other way could be make our "all activity" indexer also pickup group posts

## Rationale and Alternatives

A key success metric of this standard is to provide a minimal blueprint that gateways can build upon to ensure feature parity, in the UX goals above, across gateways. 

Prior art: 

1. near.social's NFT-based groups

> The [group's post page](https://near.org/near/widget/ComponentDetailsPage?src=mob.near/widget/N.Group&tab=source); passing to this page an NFT's contractID as the groupId query param will [render the group details](https://near.org/mob.near/widget/N.Group?groupId=v0.8bityonce.near).   A primary drawback to this NFT-based approach is that there is no canonical list of group members. Instead, group membership is inferred from the owners of a NFT per a non-socialDB API - a Mintbase API is used in the above component. 

Reviewer 2: Would that mean that every NFT contract is a group or should we put some restrictions?

Author: I think an NFT should only become a group on socialDB when an NFT owner creates a post on socialDB.  The NFT comes first, then the socialdB group space for that NFT only exists when an NFT owner writes into that social DB space.

Reviewer 2: This is a great idea 👍 the contractId could also write a post to the group so the first post ever would have to be an NFT contractId post.

> The [group's post page](https://near.org/near/widget/ComponentDetailsPage?src=mob.near/widget/N.Group&tab=source); passing to this page an NFT's contractID as the groupId query param will [render the group details](https://near.org/mob.near/widget/N.Group?groupId=v0.8bityonce.near).   A primary drawback to this NFT-based approach is that there is no canonical list of group members. Instead, group membership is inferred from the owners of a NFT per a non-socialDB API - a Mintbase API is used in the above component. 

Reviewer 2: I am also wondering how to treat NFT transfers. Should they be ignored? SO in order to become a member of a group a user should mint an NFT themselves.

Author: With group membership based on NFT ownership,  if I transfer my GroupA NFT to you,  then I no longer satisfy that criteria, lose my ability to post onto that group, and you would become a group member.

Reviewer 2: I see, so that would require indexing nft_transfer (and probably nft_burn) operations as well and update the indexer logic accordingly. Thanks!

With this approach, can we efficiently determine which group members should receive a notification when a post is created on that group's page?  How challenging would it be to ensure that only group members can post to this group's page? There is a solution here, however these implicitly defined group connections with multiple versions of group membership adds complexity. 

1. [Communities on DevHub](https://near.org/devgovgigs.near/widget/Ideas)

> DevHub's community implementation exists on a distinct contract from near.social. This results in a lack of discoverability of DevHub's groups across gateways due to its community data not being included in the indexes which power search on socialdB.

Reviewer: One workaround of this problem is implement group on contract side and define a common interface that custom indexer should implementation, a minimum example could be `create_group`, `join_group` and `leave_group` calls should be indexed. And the gateway can then rely on the indexed data to form discovery. The permission logic is still handled by custom contract

> 

Reviewer: Another alternative would be sputnik dao based group, need more thought here. A very basic thinking it needs implementation widget and custom indexer. The major benefits is advanced moderation, voting, dao capability can be introduced in the future.

Author: I shared my thoughts on a DAO-based approach [here](https://github.com/near/near-discovery/discussions/342#discussioncomment-7296473), but ultimately ended up not recommending a DAO-based approach to due to it's implementation and NEAR costs.  

With this SEP, do you see anything which would prevent a gateway operator or developer from using a DAO to managed a NFT-based group?

Reviewer: I don't think so. This approach doesn't prevent group admins to extend the NFT contract to also incorporate DAO logic.

## Specification 
Building upon the NFT-based groups implementation, we would benefit from the following the creation and maintanence of two [Core Social Indexers](https://docs.google.com/document/d/14OmYzMYX60lBm2U2Ridz6GZCeN_A6IcU_qn-yPwLTfs/edit#heading=h.nfrlp9tpavvv):

1. A dataplatform.near Core Social indexer for NFT_Groups, that tracks posts to main which include a groupContractId, e.g. writes to post.main.[groupContractId], and provides an API which lists all of the groups which have been created in socialDB. 

> 	1. To support the ability to search for groups, where search text is matched against the group's metadata such as name, symbol, description, and spec, the above indexer should fetch and store this metadata from a source such as the [Mintbase API](https://cloud.hasura.io/public/graphiql?endpoint=https%3A%2F%2Fgraph.mintbase.xyz%2Fmainnet&header=mb-api-key%3Aanon&query=query+MyQuery+%7B%0A++nft_metadata%28%0A++++where%3A+%7Breference%3A+%7B_eq%3A+%22nb0-oBR379DzoFYeYv-LesjVNmrVlDs5IqQ8hfDfnMU%22%7D%7D%0A++%29+%7B%0A++++id%0A++++media%0A++++reference_blob%0A++%7D%0A%7D%0A

Reviewer 2: I think that fetching this information within an indexer is risky and shouldn't be done in general. I think that a better approach is to index the metadata or use an existing Mintbase solution.

Author: Sounds like a fetch within an indexer is needed in either of those options. Since Indexer 1 is listening to writes to main with a groupContractId, it would have to fetch from some non-socialDB source in order to get the NFT metadata.

Reviewer 2: I am not sure if fetching from within the indexer is the right approach for the indexers in general. Maybe someone with more background in the indexers can share some thought here.

> 1. A dataplatform.near Core Social indexer for Group_Members, that maintains a list of NFT owners for each NFT_Group, and provides an API which serves a list of all AccountIds which belong to a group.

Reviewer: Nice thoughts!

Author: @milenakrawczyk perhaps we can use mintbase and avoid creating this second indexer.

> 1. A dataplatform.near Core Social indexer for Group_Members, that maintains a list of NFT owners for each NFT_Group, and provides an API which serves a list of all AccountIds which belong to a group.

Reviewer 2: I'm wondering how a group membership should be tracked. We will index posts so one natural way is to see if a post in the group was created. Is there any other way that this could be achieved?

Author: [Indexer 1](https://github.com/NearSocial/SEPs/pull/2/files#diff-d728a3d80ccb3924ed1332507ef858a82a423e70d7faccc84ff297b0744c8a32R61) has been suggested as the way to track group membership.  Are you concerned that this is not appropriate?

Reviewer 2: That was an old comment of mine and I think that we may use one of the indexers listed in this proposal to fetch the members. We may index nft_mint operations to determine who is a member of a group and combine this with a condition that every valid group should have one post posted by the owner.

## Drawbacks

> - The NFT-based groups implementation shifts the responsibility of valid NFT management onto users. However, gateways can mitigate this with a Mintbase integration or the ContractWizardUI to faciliate the creation of NFTs.

Reviewer: I have a question: is there a good practice to push this common logic to all gateways, including near.org/near.social and 3rd party gateways?

Author: I think we may want to advocate that gateways use a specific component that facilitates the creation of an NFT, we could start with [this one](https://near.org/contractwizard.near/widget/ContractWizardUI)

Reviewer: Got it. Thanks!

- The ability to forcibly remove someone from a group, e.g. for violating group guidelines, might not exist - depending on the functionality of the NFT contract for a given group. A potential workaround would be to 'moderate' the user by blocking their posts or banning their account - this would be gateway specific. A gateway could also offer a 'recommended' path for creating a group that includes the creation of an NFT collection with a contract that includes rules that allows the group creator to initiate a transfer to allow the removal of the NFT from any holder.

> 

Reviewer 2: If there is an ability to remove a user from a group, how should this information be stored in the SocialDB? The indexer needs to know if a group member was removed.

Author: @milenakrawczyk what do you recommend?  The schema that the indexer populates could minimally include a key/value pairing of `NFTContractID:[AccountIdList]`

Reviewer 2: Could you explain further? Where would this key/pair be stored?
If we base our approach on NFTs, then we would need to combine the data from NFTs with the key/value pairs you mentioned.

Reviewer 2: Maybe removing a user from a group could be achieved by expiring an NFT as was mentioned by @evgenykuzyakov. We could index specific calls to know if a user is still a member or not.

Author: NFTs are not required to have an expiration, it's an optional field. So we cannot depend on leveraging expiration for moderation.

Author: we need to determine this as part of the POC implementation
> Where would this key/pair be stored?

Where each group object contains the following required fields: 

- a unique identifier: String
- originating gateway: String
- members: [AccountList](https://github.com/NearSocial/standards/blob/main/types/common/AccountList.md)
- metadata: [Metadata](https://github.com/NearSocial/standards/blob/main/types/common/Metadata.md?plain=1)

`{
	"alice.near": {...},
	"mob.near": {...},
	"groups" : [
		{
			"1290387aslkdjf09182": {
				"gateway": "near.org",
				"members": {
					"alice.near": "",
					"bob.near": "",
				},
				metadata: {
				"name": "zksync fans"
					...
				},
			}
		},
		{
			"00110357asiedjf0935t": {
				"gateway": "near.social",
				"members": {
					"gauthamravi.near": "",
					"mob.near": "",
					"root.near": "",
				},
				metadata: {
					"name": "thebestgroup", 
					...
				},

			}
		}
		...
	]
}`

This alternative is suboptimal because 
- It breaks the socialDB pattern of top level objects being Accounts
- Restricting non-group members from writing to a group's namespace would require an extension of socialdB's permissions management; increasing its complexity. 

> - It would require more implementation time and cross-ecosystem coordination.

Reviewer: I agree with the these rationales.

---
SEP: 2
Title: WIP Standardize Groups
Authors: Charles G. <charles@near.org>
Status: Draft
DiscussionsTo: https://github.com/nearsocial/SEPs/pull/2
Version: 1.0.0
Created: 2023-10-15
Last Updated: 2023-10-15
---

## Summary

This document serves to come to an agreement on a specification for a new "Groups" experience on Near Social, designed to allow people to create and discover communities around specific topics, interests, projects, governance, etc.

## Motivation

We need a SocialDB standard for groups to support discoverability on near.org and interoperability between gateways and dApps to mitigate the [compatibility problems](https://near.social/devgovgigs.near/widget/gigs-board.pages.Post?id=1034) which arise when dApps build distinct implementations of communities.

## UX Goals

> 1. To allow a group of users to moderate with sovernity a scope of socialDB limited to the members of the group. e.g. group members have read + write access to the group's socialDB namespace. 

Reviewer 3: ```suggestion
1. To allow a group of users to moderate with sovereignty a scope of socialDB limited to the members of the group. e.g. group members have read + write access to the group's socialDB namespace. 
```

1. Non-members only have read-only access to view group content
1. Group members should receive notifications for all relevant social interactions in a group (e.g. Like, Reply for Posts and Discussions)
1. All content posted to groups that a user is part of should appear the user’s "Following" activity feed. 
1. Leaving a group will stop content from appearing on a user’s feed.

> 1. Group content should only appear in the “All” activity feed if a group member reposts a group post to the main feed.

Reviewer 4: @charleslavon This is not an action that is currently supported today on near.org but it does exist on near.social. Is the idea that a user will be able to select where they can repost content? Is the idea here that a user would be able to select where they can repost content? This is probably out of scope for this discussion around groups so feel free to resolve and we can discuss at a later time.

1. One should be able to retrieve a list of all groups in socialDB
1. One should be able to retrieve a list of all users in a group.

> 1. One should be able to search for groups, where search text is matched against the group's name and metadata.

Reviewer 4: @charleslavon I think we should also call out a general UX goal related to content moderation and member moderation in groups. I think those functions would cut across different gateway experiences

Author: @ravigautham if you were to write that one or multiple feature requirements, how would you describe it?

Prior art: 

1. near.social's NFT-based groups

The [group's post page](https://near.org/near/widget/ComponentDetailsPage?src=mob.near/widget/N.Group&tab=source); passing to this page an NFT's contractID as the groupId query param will [render the group details](https://near.org/mob.near/widget/N.Group?groupId=v0.8bityonce.near).   A primary drawback to this NFT-based approach is that there is no canonical list of group members. Instead, group membership is inferred from the owners of a NFT per a non-socialDB API - a Mintbase API is used in the above component. Depending on which API server is used to fetch the owners of a given NFT, this approach can lead to erroneous representations of who is a member of a group. For details, see this [Mintbase API Issue](https://github.com/Mintbase/mintbase-js/issues/412) and this [Enhanced API issue](https://github.com/near/near-enhanced-api-server/issues/70). 

1. devs.near's groups [defined implicitly by follow-style graph connections](https://near.org/devs.near/widget/every.group). 

> A primary drawback to this graph-based approach is that there is no canonical source of group membership. A single group of N members can have M distinct versions of who belongs in that group, where M is the number of group members who decline to "mutually follow" each newly requested group member.  Group membership is managed by an object on the namespace of anyone who has requested access to a group, which contains a group identifer as a key with a value of a list of accountIds (AccountList). Even if no one approves someone's request to join a group, the requestor would retain a valid instance of that groupIdentifer:AccountList on their socialDB namespace - with their accountID included as a group member. This results in muliple versions of graphs which represent the definition of members of a group. For more details on this implementation see minute 33:45 [here](https://youtu.be/HAyweTF7IjU?si=z_hfct3oK_86nt_o)

Reviewer 3: ```suggestion
A primary drawback to this graph-based approach is that there is no canonical source of group membership. A single group of N members can have M distinct versions of who belongs in that group, where M is the number of group members who decline to "mutually follow" each newly requested group member.  Group membership is managed by an object on the namespace of anyone who has requested access to a group, which contains a group identifier as a key with a value of a list of accountIds (AccountList). Even if no one approves someone's request to join a group, the requestor will retain a valid instance of that groupIdentifer:AccountList on their socialDB namespace - with their accountID included as a group member. This results in muliple versions of graphs which represent the definition of members of a group. For more details on this implementation, see minute 33:45 [here](https://youtu.be/HAyweTF7IjU?si=z_hfct3oK_86nt_o)
```

> Which this approach, can we efficiently determine which group members should receive a notification when a post is created on that group's page?  How challenging would it be to ensure that only group members can posts to this group's page? There is a solution here, however these implicitly defined group connections with multiple versions of group membership adds complexity. 

Reviewer 5: Excellent proposal @charleslavon !! I think you've done a great job thoroughly assessing all of the already attempted possibilities. Thank you for this 😄

Reviewer 4: +1 👏👏👏 @charleslavon

Author: thanks guys

> Which this approach, can we efficiently determine which group members should receive a notification when a post is created on that group's page?  How challenging would it be to ensure that only group members can posts to this group's page? There is a solution here, however these implicitly defined group connections with multiple versions of group membership adds complexity. 

Reviewer 3: ```suggestion
With this approach, can we efficiently determine which group members should receive a notification when a post is created on that group's page?  How challenging would it be to ensure that only group members can post to this group's page? There is a solution here, however these implicitly defined group connections with multiple versions of group membership adds complexity. 
```

1. [Communities on DevHub](https://near.org/devgovgigs.near/widget/Ideas)

> DevHub's community implementation exists on a distint contract from near.social. This results in a lack of discoverability of DevHub's groups across gateways due to its community data not being included in the indexes which power search on socialdB.

Reviewer 3: ```suggestion
DevHub's community implementation exists on a distinct contract from near.social. This results in a lack of discoverability of DevHub's groups across gateways due to its community data not being included in the indexes which power search on socialdB.
```

Reviewer 3: I agree that this should live on social-db for simplicity & discoverability.

## Specification 

> 1. Adjacent to root level socialDB accounts, create a new top account called `groups.near`. However, since this account name has already been claimed, I have slightly rearranged the same letters to use the spanish translation: `grupos.near`. For convience, gateway operators could add a redirect from `gatway.com/near/groups.near` -> `gatway.com/near/grupos.near` if desired.

Reviewer 3: ```suggestion
1. Adjacent to root level socialDB accounts, create a new top account called `groups.near`. However, since this account name has already been claimed, I have slightly rearranged the same letters to use the spanish translation: `grupos.near`. For convenience, gateway operators could add a redirect from `gateway.com/near/groups.near` -> `gateway.com/near/grupos.near` if desired.
```
What about using `groups` as a root level like `aurora`?

Author: @thisisjoshford please elaborate

Reviewer 3: I just mean we have TLA accounts like `aurora`, `near`, etc. and could use simply `groups` over `groups.near`

Author: I suspect that it would be difficult to justify the introduction of `groups` as a TLA because we don't entirely need unique accounts for groups. Also this would introduce the problem (benefit?) of determining how to deal with unique group names within this domain. Would you recommend `[groupName].groups`? Then we would need to deal with squatters sitting on cool group names like `theopenweb.groups`.  Alternatively we could using unique identifiers like the groupUID mentioend in this SEP, but I think `00110357asiedjf0935t.groups` is unsightly.

2. Represent a group via two data structures. A list of group members of type [AccountList](https://github.com/NearSocial/standards/blob/main/types/common/AccountList.md), and a new object standard `Group`, which consists of the following required fields: 

- a unique identifier: String (randomly generated unique ID i.e. [generateUID()](https://near.org/near/widget/ComponentDetailsPage?src=devs.near/widget/group.create&tab=source)) 
- gateway: String (the gateway where the group was initially created)
- metadata: [Metadata](https://github.com/NearSocial/standards/blob/main/types/common/Metadata.md?plain=1)

> 	Seperating Group data into distinct objects, `grupos.near/post/index/[groupUID]_members`and `grupos.near/post/index/[groupUID]_data`, affords using socialDB's namespace permissions to restrict write-access of a Group's metadata to specific users, i.e. groups admins or the creator of the group, while allowing all group members to write to the list of members.

Reviewer 5: I think utilizing namespaces like this is a really creative solution

Question though; why this:
```
`grupos.near/post/index/[groupUID]_members`
`grupos.near/post/index/[groupUID]_data`
```
instead of:

```
grupos.near/post/index/[groupUID]/members
grupos.near/post/index/[groupUID]/data
```

And how will a group member post to a group's feed? 
Are you envisioning individual group "room" permissions, such as:

```
grupos.near/post/index/[groupUID]/room // permissions to post in all rooms

grupos.near/post/index/[groupUID]/room/general // individual room permissions 
grupos.near/post/index/[groupUID]/room/random
```

or will any user have permissions to post:
```
efiz.near/post/index/[groupId]
```
And the feed will be filtered by only the members in the member list?

(or some other way?)

Author: hey @elliotBraem, thanks for the thoughtful feedback and questions. 
```
grupos.near/post/index/[groupUID]/members
grupos.near/post/index/[groupUID]/data
```
I liked this approach as well, but form my tests, social.near's grant_write_permissions are not granular enough to limit write access to a space within a given index. e.g. By granting permissions to the `groupUID` index, one would have access to that entire space.

Author: @elliotBraem I've addressed your questions re:posts [here](https://github.com/NearSocial/SEPs/pull/2#discussion_r1362916231). Let me know if I've missed something

3. With the above we can agree to use the namespace @ `grupos.near/post/index` to contain groups:

{
    "alice.near": {...},
    "mob.near": {...},
    "grupos.near":
    {
        "post":
        {
            "index":
            {
                "1290387aslkdjf09182_members":
                {
                    "alice.near": "",
					"bob.near": "",
                },
                "1290387aslkdjf09182_data":
                {
                    "name": "zksync fans",
                    "gateway": "near.org",
                    "metadata":
                    {
                        "description": "we love zksync",
                        "tags": ["ether-js", "eth", "multichain"]
                    }
                },
                "00110357asiedjf0935t_members":
                {
                    "gauthamravi.near": "",
                    "mob.near": "",
                    "root.near": ""
                },
                 "00110357asiedjf0935t_data":
                {
                    "name": "the best group",

>                     "gateway": "near.social",

Reviewer 3: ```suggestion
                    "gateway": "near.social",
                    "require-nft": true,
                    "nft-id": "group-nft.bestgroup.near"
```

Reviewer 3: I like @evgenykuzyakov's NFT use for group membership and wonder if we could have a hybrid approach here where creators can choose if owning an NFT is required for membership.

Author: @thisisjoshford how might you suggest we address the challenge in determining NFT ownership [here](https://github.com/NearSocial/SEPs/pull/2/files#diff-d728a3d80ccb3924ed1332507ef858a82a423e70d7faccc84ff297b0744c8a32R43)?

Reviewer 4: @charleslavon @thisisjoshford  Would the only solution here be to create a standardized NFT API for SocialDB? That way any app or gateway built on the SocialDB smart contract would be required to use this specific API to obtain a standardized list of NFT owners that would be the same across different experiences.

 Is NFT ownership to gate membership something that can be a future enhancement to this standard?

Author: @ravigautham a gateway operator could add NFT-ownership requirement without it needing to be included in this standard.

                    "metadata":
                    {
                        "description": "super group",

>                         "guidelines": "be excellent humans."

Reviewer 3: 💜

                    }
                }
            }
        }
    }
}

### Supporting this requires: 

1. A common goods provider to maintain a service that abstracts and securely exposes calls to social.near's `grant_write_access` and `set` methods via an API such as
	/groups/[groupUID]/add/[accountId]  (adds a user to a group and creates the group if it doesn't exist)
	/groups/[groupUID]/remove/[accountId] 
	/groups/[groupUID]/verify?accountIds=[accountIds] (a convenience method to expose the canonical source of truth of group membership. Returns an object of [accountId]:boolean pairs to confirm which accountIds have write access fo the `_members` scope of the given groupUID. This API essentially wraps calls to `is_write_permission_granted`, modified to accept multiple AccountIds.)
2. A common goods provider to maintain an indexer and service to fan-out group notifications into notifications sent to each group's member.

## Drawbacks

> - By granting each group member write access to `grupos.near/post/index/[groupUID]_members`, any group member could leverage near.social's `set` to remove others from the list of group members. However, this is of minimal impact since this list is independent of the management of write permissions. In this case, this means that a removed group member would retain their ablity to write the list of members. Easily supporting data recovery from a previous block.  

Reviewer 4: @charleslavon  How would this effect scenarios where group members actually would need to be removed due to behavior that goes against a group's guidelines? Would they need to be removed from both the member list and write permissions list?

Author: That's correct. A gateway operator would need to couple the mutation of the write permissions and the member AccountList on the UI affordance which adds/removes group members.

> - By granting each group member write access to `grupos.near/post/index/[groupUID]_members`, any group member could leverage near.social's `set` to remove others from the list of group members. However, this is of minimal impact since this list is independent of the management of write permissions. In this case, this means that a removed group member would retain their ablity to write the list of members. Easily supporting data recovery from a previous block.  

Reviewer 5: Do we need to grant each group member write access to it? It doesn't seem like a requirement to me, or am I missing an implication?

Reviewer 4: Granting access would be the only way for group members to post or write content to the group page correct?

Reviewer 3: From what I'm following here, yes. You can have both read & write access, only ready access, or neither.

Author: Great question @elliotBraem.  Granting write access to each group member gives the group complete control over managing the member list; allowing each group member to remove/add people at will.  If we do not grant write access to `_members`, this control would need to live at the gateway or perhaps with the creator of the group. We could also consider granting write access to a group role like `Admin`.

Regarding how to create posts, they would need to live in another index. Perhaps `grupos.near/post/index/[groupUID]_main`. By this means we can limit post creation to group members. 

In addition to the two required indexes, `_members` and `_data`,  should gateways facilitate the creation of an arbitrary number of indexes per group?

Author: @thisisjoshford I would amend that to say that everyone has read access, only group members have read & write access.

Reviewer 4: @charleslavon I think granting write access to a specific group role like the creator of the group who will automatically be the first admiin and any other designated admins. I'm wary about giving write access to all group members as this could facilitate unintended use of the member removal feature. Even though those members would still have the ability to post content, it would result in a confusing user experience. 

I think this primitive should allow for the creation of multiple different group roles and different sets of permissions for each of those roles. We might leverage the admin/member structure on near.org, but another gateway might want to add a third or fourth role to their implementation of groups. How do we ensure that this primitive is flexible enough to allow for that?

`{
	"alice.near": {...},
	"mob.near": {...},
	"groups" : [
		{
			"1290387aslkdjf09182": {
				"gateway": "near.org",
				"members": {
					"alice.near": "",
					"bob.near": "",
				},
				metadata: {
				"name": "zksync fans"
					...
				},
			}
		},
		{
			"00110357asiedjf0935t": {
				"gateway": "near.social",
				"members": {
					"gauthamravi.near": "",
					"mob.near": "",
					"root.near": "",
				},
				metadata: {
					"name": "thebestgroup", 
					...
				},

			}
		}
		...
	]
}`

This alternative is suboptimal because 
- It breaks the socialDB pattern of top level objects being Accounts
- Restricting non-group members from writing to a group's namespace would require an extension of socialdB's permissions management; increasing its complexity. 
- It would require more implementation time and cross-ecosystem coordination.

# Potential Impact and Dependencies
//TBD
Here we aim to be mindful of our environment and generate empathy towards others who may be impacted by our decisions.

- *What other systems or teams are affected by this proposal?*
- *How could this be exploited by malicious attackers?*

# Unresolved questions

> - How might a group owner/admin or member deal with malicous group members?

Reviewer 3: cc @tidus974
